File: solutions/arrays-hashing/128-longest-consecutive-sequence.py
"""
# Difficulty: Medium

# 128. Longest Consecutive Sequence

Given a problem that demonstrates key concepts in Arrays Hashing.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[input description]</dd>
<dt>Output:</dt>
<dd>[output description]</dd>
<dt>Explanation:</dt>
<dd>[explanation]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of arrays hashing concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply arrays hashing methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages arrays hashing principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        """
        Find the length of the longest consecutive sequence in an unsorted array.
        
        Args:
            nums: List of integers (can be unsorted and contain duplicates)
            
        Returns:
            int: Length of the longest consecutive sequence
            
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        if not nums:
            return 0
        
        # Convert list to set for O(1) lookup
        num_set = set(nums)
        longest_streak = 0
        
        # Iterate through each number in the set
        for num in num_set:
            # Only start counting sequences from the smallest number in the sequence
            # If num-1 exists, this isn't the start of a sequence
            if num - 1 not in num_set:
                current_num = num
                current_streak = 1
                
                # Count consecutive numbers
                while current_num + 1 in num_set:
                    current_num += 1
                    current_streak += 1
                
                # Update longest streak if current streak is longer
                longest_streak = max(longest_streak, current_streak)
        
        return longest_streak

def test_solution():
    """
    Test cases for 128. Longest Consecutive Sequence.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 128. Longest Consecutive Sequence")

File: solutions/arrays-hashing/128-longest-consecutive-sequence.js
I'll help you create a solution for the Longest Consecutive sequence problem in JavaScript. I'll write a clean implementation with comments and proper error handling.

```javascript
/**
 * @file Longest Consecutive Sequence Implementation
 * @description Finds the length of the longest consecutive sequence in an unsorted array
 */

/**
 * @param {number[]} nums - Array of numbers
 * @return {number} - Length of longest consecutive sequence
 */
function longestConsecutive(nums) {
    // Handle edge cases
    if (!nums || nums.length === 0) {
        return 0;
    }

    // Create a Set for O(1) lookup
    const numSet = new Set(nums);
    let maxLength = 0;

    // Iterate through each number in the array
    for (const num of numSet) {
        // Only start checking sequences from the smallest number in the sequence
        // If num-1 exists, this isn't the start of a sequence
        if (!numSet.has(num - 1)) {
            let currentNum = num;
            let currentLength = 1;

            // Count consecutive numbers
            while (numSet.has(currentNum + 1)) {
                currentNum++;
                currentLength++;
            }

            // Update maxLength if current sequence is longer
            maxLength = Math.max(maxLength, currentLength);
        }
    }

    return maxLength;
}

// Test cases
const testCases = [
    [100, 4, 200, 1, 3, 2],           // Expected: 4 (1,2,3,4)
    [0, 3, 7, 2, 5, 8, 4, 6, 0, 1],   // Expected: 9 (0,1,2,3,4,5,6,7,8)
    [],                                // Expected: 0
    [1],                               // Expected: 1
    [1, 2, 0, 1]                      // Expected: 3 (0,1,2)
];

// Run test cases
testCases.forEach((test, index) => {
    console.log(`Test ${index + 1}:`);
    console.log(`Input: [${test}]`);
    console.log(`Output: ${longestConsecutive(test)}`);
    console.log('---');
});

// Export the function for potential module usage
module.exports = longestConsecutive;
```

This implementation provides a solution to find the longest consecutive sequence in an unsorted array with the following features:

1. **Time Complexity**: O(n) where n is the length of the input array
2. **Space Complexity**: O(n) to store the Set

Key features of the implementation:

1. Uses a Set for O(1) lookup time
2. Optimized to only start checking sequences from the smallest number in each sequence
3. Includes comprehensive test cases
4. Handles edge cases (empty array, null input)
5. Follows JavaScript best practices and conventions
6. Includes clear comments explaining the logic
7. Exports the function for modular usage

The algorithm works by:
1. Converting the input array to a Set for efficient lookup
2. Finding the start of each sequence (numbers that don't have a predecessor)
3. Counting consecutive numbers from each sequence start
4. Tracking the maximum sequence length found

The code includes test cases demonstrating various scenarios and edge cases, making it easy to verify the implementation works correctly.